cardinal/gamestate/map.go

```diff
@@ -0,0 +1,272 @@
+package gamestate
+
+import (
+	"context"
+
+	"github.com/rotisserie/eris"
+)
+
+type MapStorage[K comparable, V any] struct {
+	internalMap map[K]V
+}
+
+func NewMapStorage[K comparable, V any]() *MapStorage[K, V] {
+	return &MapStorage[K, V]{
+		internalMap: make(map[K]V),
+	}
+}
+
+func (m *MapStorage[K, V]) Keys(_ context.Context) ([]K, error) {
+	res := make([]K, 0)
+	for key := range m.internalMap {
+		res = append(res, key)
+	}
+	return res, nil
+}
+
+func (m *MapStorage[K, V]) Get(_ context.Context, key K) (any, error) {
+	var resAny any
+	res, ok := m.internalMap[key]
+	if !ok {
+		return nil, eris.New("key does not exist in storage")
+	}
+	resAny = res
+	return resAny, nil
+}
+
+func (m *MapStorage[K, V]) Clear(_ context.Context) error {
+	m.internalMap = make(map[K]V)
+	return nil
+}
+
+func (m *MapStorage[K, V]) GetFloat64(_ context.Context, key K) (float64, error) {
+	var res float64
+	var resAny any
+	resRaw, ok := m.internalMap[key]
+	if !ok {
+		return 0, eris.New("does not exist in map")
+	}
+	resAny = resRaw
+	res, ok = resAny.(float64)
+	if !ok {
+		return 0, eris.New("cannot convert to float")
+	}
+	return res, nil
+}
+func (m *MapStorage[K, V]) GetFloat32(_ context.Context, key K) (float32, error) {
+	var res float32
+	var resAny any
+	resRaw, ok := m.internalMap[key]
+	if !ok {
+		return 0, eris.New("does not exist in map")
+	}
+	resAny = resRaw
+	res, ok = resAny.(float32)
+	if !ok {
+		return 0, eris.New("cannot convert to float")
```

Comment:
Maybe we should specific float32 in this error msg, given we also have a float64?